### DISCUSSION TITLE
env-vars behaves as build-args by design.

### DISCUSSION THREAD

**Discussion by skrhlm** [vercel/next.js#22408] [Help]
**What version of Next.js are you using?**

10.0.7

**What version of Node.js are you using?**

14.x

**What browser are you using?**

Chrome

**What operating system are you using?**

MacOs

**How are you deploying your application?**

Kubernetes

**Describe the Bug**

Simply put the `env-vars` act as build-args, it is not possible to do any override of env-variables in runtime.

**Expected Behavior**

Env-vars ( when not using export ) should not be bundled, but updatable through the environment in runtime.

**To Reproduce**

I've made a repository showcasing the frustration
https://github.com/skrhlm/NEXT-JS-ENV-VARS

**Comment by LuudJanssen** [comment]
Hi @skrhlm, I think you're looking for [Runtime Configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration). This allows you to pass environment variables (through `process.env`) to runtime configuration. The reason Next.js bundles environment variables during build time by default is explained at the start of the Runtime Configuration documentation:

> Generally you'll want to use build-time environment variables to provide your configuration. The reason for this is that runtime configuration adds rendering / initialization overhead and is incompatible with Automatic Static Optimization.

**Comment by skrhlm** [comment]
Hello @LuudJanssen ! 

Thanks for your answer. You are right that that using runtime configuration does indeed work. However, I made [a new example](https://github.com/skrhlm/NEXT-JS-ENV-VARS/tree/app-get-initial-props), this time using `getInitialProps`in `_app.js` to force pages out of automatic static optimization.

This works for all examples, but the one using getStaticProps. Which makes sense since you're [unable to opt-out of static optimization when using getStaticProps ](https://github.com/vercel/next.js/blob/master/errors/opt-out-auto-static-optimization.md).

However, the fact still stands that there should be a way to override the default behaviour for getStaticProps. There is still no way to get a site-wide variable through. I understand that this would add some overhead, but a simple bootstrapping process on startup should be able to handle this reasonably efficient.

**Reply by LuudJanssen** [reply]
Hi @skrhlm! The problem is that static pages are created and rendered on build time. This means your variable should also be available during the build.

I understand that you want to build a Docker container once and inject it with different environment variables to alter the behavior, but if the resulting Docker build should contain statically rendered pages, there's just no way around it.

Let me explain the problem. Say you have an environment variable called `VERSION_NUMBER`. You use `getStaticProps` to statically render a page and you pass `VERSION_NUMBER` through `getStaticProps` and you display it in the render function.

Now you run `next build`. During this time, the build server needs to render the entire HTML for the static page, to serve it statically. Since the render function has a dependency on `VERSION_NUMBER` it needs to be known at build time.

If you want to change `VERSION_NUMBER` later, Next.js offers `getServerSideProps` [or `getStaticProps` with revalidation after a certain amount of time](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration).

Other options are:
- Fetching the environment variable through the client
- Running the `next build` on the startup of the Docker container with the injected variables. _Very bad idea, please don't do this_.

**Comment by michrome** [comment]
Noting a [related discussion with an example using `serverRuntimeConfig`](https://github.com/vercel/next.js/discussions/22243#discussioncomment-382394).

**Comment by chrisj-back2work** [comment]
Hi @skrhlm  -- your demo repo isn't accessible. Is that in a public location somewhere?
